### ISSUE TITLE
[Enhancement]: Force transcode for downloaded file

### ISSUE THREAD

**Issue by mew1033** [advplyr/audiobookshelf#2477]
### Describe the feature/enhancement

I'd like to be able to optionally force a transcode of a file before downloading it for offline playback. Maybe the file is too big, or it's in a format my device doesn't recognize. Regardless, it'd be nice to be able to request an offline download of a specific type.

**Comment by Darkblizzard21** [comment]
Im currently facing a simmilar issue. I have a audiobook which spans over 98h and therefore is almost 5.5GB big M4B file.
When i try to play it back on mobile the server takes long to responde (because of the transcoding) so it would be easier to just play it locally, but android is not able to play it back.

**Comment by advplyr** [comment]
@Darkblizzard21 I recommend re-encoding that M4B file because they should be able to play without requiring a transcode.

**Comment by Darkblizzard21** [comment]
Would probalby be a nice feature regardless. Also if audiobookshelf is able to live transcode the file it should not be a big effort to implement encoding as part of the tools or as part of the download process to mobile (but i can imagin that it is way more complacted than i assume).

Now the OffTopic Part:
I already tried to transcode the file using ffmpeg but all my settings failed so i now probalb will just stick to playing it live transcoded while on wifi.

**Comment by advplyr** [comment]
Yeah I agree it would be a nice feature. In your case it seems like you have a poorly encoded M4B file (or something with your device) so I still recommend fixing that. It could be as simple as `ffmpeg -i input.m4b -codec:a aac output.m4b`

**Comment by mew1033** [comment]
Android should definitely support m4b files. I wonder what's going on with that file? Maybe it's not getting downloaded correctly because it's so big?

**Comment by RaHoni** [comment]
@advplyr do you have a preference for the API for this?

Otherwise I would recommend something like 
`api/items/:id/file/:fileid/download?containerformats=mp3,m4b&audiocodecs=aac,mp3`

Which would check if it is already available in an accepted format. And if not it gets converted before the download starts. 

As optimization we could test if for example only the container needs to be changed and than use '-codec:a copy'.

**Comment by delize** [comment]
# [Enhancement] Selectable download format (transcode-on-download) — keep hi-fi originals, download a mobile-playable copy

### Use case

Some of my audiobooks are encoded as **5.1-channel Dolby Digital Plus + Dolby Atmos** (E-AC-3, tag `ec-3`, 768 kbps, 48 kHz, 6ch). They **stream perfectly** through the server (the web/desktop player transcodes them), but on the **Android app** they **download and then fail at "Download complete. Processing…"** — the local library item never gets created, and even with the bytes present the file won't decode on-device: mobile `MediaCodec` generally has no 5.1 E-AC-3/Atmos decoder, and ExoPlayer/Media3 has no software fallback.

Notably, **nothing appears in the server Logs tab** — because the server *succeeds*. It streams the whole file (HTTP 200, full byte count). The failure is entirely in the client's post-download processing, so the server has nothing to log. That makes this hard to diagnose from the server side.

### What I'd like

A **per-download format choice** (the way Navidrome clients like Tempus do it):

- **Original / no transcode (default)** — download the file byte-for-byte, for archival or players that can handle it.
- **Transcode to a compatible target** — e.g. AAC or Opus at a selectable bitrate (mono for spoken word) — so the *same* book can be downloaded in a form that actually plays on the phone, **without altering the high-quality original stored on the server**.

That combination is the goal: hi-fi originals stay for home/streaming; mobile downloads a playable, far smaller copy on demand (a 6.0 GB 5.1 E-AC-3 book → ~0.5 GB AAC mono).

### Repro

- Server: audiobookshelf `latest` (Docker). App: Android `v0.12.0-beta`, device Samsung SM-X820 (Android / SDK 36).
- Add a book whose audio is E-AC-3 5.1 (**sample attached**).
- Stream it in the web player → works.
- Download it in the Android app → progresses to **"Download complete. Processing…"**, then fails to create the local library item / won't play. Server Logs show a clean, successful download.

### Sample / diagnostics (attached)

- [ffprobe-real-file.txt](https://github.com/user-attachments/files/30677009/ffprobe-real-file.txt) - ffprobe of a real affected file.
- [abs-eac3-samples.zip](https://github.com/user-attachments/files/30677011/abs-eac3-samples.zip) - 30 s **synthetic** 5.1 E-AC-3 @ 768 kbps in an MP4/M4B container (no copyrighted audio); reproduces the exact codec condition.

(If you want a real sample of the Audiobook, please let me know and can provide a 30 second clip)



```
codec_name=eac3   profile=Dolby Digital Plus + Dolby Atmos   codec_tag=ec-3
channels=6   channel_layout=5.1(side)   48000 Hz   768 kbps
container: mov,mp4,m4a   duration ≈ 17.4 h   size ≈ 6.0 GB
```

### Related

- advplyr/audiobookshelf#2477 — Force transcode for downloaded file
- advplyr/audiobookshelf-app#1875 — Transcode for downloads (mobile)
- advplyr/audiobookshelf-app#1741 — eac3 files not playing on Android
- advplyr/audiobookshelf-app#1714 — streams fine, fails when downloaded
- advplyr/audiobookshelf#4236 — xHE-AAC support

**Comment by delize** [comment]
@Vito0912 

Can you provide feedback about the downvote? Cause I am having an actual issue with this with my audiobooks, and instead of opening a NEW issue, found this one. 

However, it seems like it is an obvious issue that can be solved either: 

1. Server side
2. Client side

In the referenced issues above, the "solutions" on the client side, have been "transcode your audiobook to mp3" as in 1714, or "the Web Client only recently received support" in 1741.  I don't want to transcode the the audiobook down to mp3. But even going down the rabbit hole to other linked issues: https://github.com/GrakovNe/lissen-android/issues/321 - it could be possible.

On the server side, it would be providing a transcode that is then downloaded to the client - so the original on the web would still be the full bitrate. 

In either case, if the solutio is "use Lissen", Lissen unfortunately, doesn't seem to support SSO (even though on the roadmap). 

So the other alternative would be to support EAC on the client side, but not all clients will be able to support it (given that the codecs don't exist):
 
- https://github.com/shane9b3/LitLyric---Beta/issues/39
- https://github.com/advplyr/audiobookshelf-app/issues/1741

So happy to hear feedback about this, as opposed to just a downvote. Unless, you just don't like the AI backed posting, which could have posted a better except even with research being done.

**Comment by Vito0912** [comment]
No, some points are valid. It's just plain AI slop. Having the option to transcode a download would be nice. But the issue for everything "you" stated exists. It's just AI noise.

And ofc Lissen does support OIDC/SSO. But, no, I don't think the solution is Lissen ofc. It should be handled by the first party apps.

And the long answer for a simple downvote for an obvious AI generated answer also speaks for itself.

Edit: (But ofc this is just my opinion and I just added a simple thumb down :). If other think that these AI messages, which flood the issues, provide real value you have any right to do so.

**Comment by delize** [comment]
> And ofc Lissen does support OIDC/SSO.

Interesting, because, on default login screen (only shows user/pass combo - when not filled out), looking at the readme, and looking in Lissen's settings, there is no reference for SSO/OIDC support. 

But will give that a try in that case instead.

Admtitedly, I didn't do a ton of searching, in the middle of trying to pack for a trip, and have spent more than 30 minutes trying to download tracks that would then fail while I pack and then come to back to the device only for them to not appear. So it is a combo of frustration for something that just works when streaming, but doesn't with no indication as to why when downloading.

**Comment by mew1033** [comment]
I just wanted to pop in here and say that I still want this feature. 😁

In my case, it's more about storage. I'd like to keep full FLAC files on the server, and download mp3s (or something more compact) on my phone. As a stopgap, I wouldn't mind doing the transcoding myself and adding the files to the book folder. The client could (gated behind an advanced option perhaps?) decide which files it wants.
